src/nl/hannahsten/texifyidea/index/LatexProjectStructure.kt

```diff
@@ -0,0 +1,687 @@
+            inputs.filterIsInstance<String>().mapNotNull {
+                tomlFile.parent?.findFile("src/$it")
+            }.toSet()
+        }
+        return filesets
+    }
+
+    private fun buildFilesets(project: Project): LatexProjectFilesets {
+        val startTime = System.currentTimeMillis()
+        val projectInfo = makePreparation(project)
+        val roots = getPossibleRootFiles(project)
+        val allFilesetInfo = mutableListOf<FilesetInfo>()
+        val processedFiles = mutableSetOf<VirtualFile>()
+
+        for (root in roots) {
+            val fileset = buildFilesetFromRoot(root, project, projectInfo)
+            allFilesetInfo.add(fileset)
+            processedFiles.addAll(fileset.files)
+        }
+        // let us try to merge the filesets that are created by Tectonic.toml files
+        val tectonicFilesets = findTectonicTomlInclusions(project)
+        for (files in tectonicFilesets) {
+            if (files.isEmpty()) continue
+            for (fs in allFilesetInfo) {
+                if (fs.root in files) {
+                    // this fileset already contains the root file, we can merge the files
+                    mergeFilesIntoFileset(files, fs)
+                }
+            }
+        }
+//      deal with the remaining files, or we can just ignore them?
+        val projectFileIndex = ProjectFileIndex.getInstance(project)
+        projectFileIndex.iterateContent { vf ->
+            if (vf.fileType == LatexFileType && vf !in processedFiles) {
+                // this is a standalone file, not included in any fileset yet
+                val fileset = buildFilesetFromRoot(vf, project, projectInfo)
+                allFilesetInfo.add(fileset)
+                processedFiles.addAll(fileset.files)
+            }
+            true
+        }
+
+        // final processing
+        val allFilesets = allFilesetInfo.map { Fileset(it.files, it.root) }.toSet()
+        val mapping: Map<VirtualFile, MutableSet<Fileset>> = buildMap {
+            allFilesets.forEach {
+                it.files.forEach { file ->
+                    getOrPut(file) { mutableSetOf() }.add(it)
+                }
+            }
+        }
+
+        val dataMapping = mapping.mapValues { (file, filesets) ->
+            val allFiles = filesets.flatMapTo(mutableSetOf(file)) { it.files }
+            val scope = GlobalSearchScope.filesWithLibrariesScope(project, allFiles)
+            FilesetData(filesets, allFiles, scope)
+        }
+
+        val elapsedTime = System.currentTimeMillis() - startTime
+        totalBuildTime.addAndGet(elapsedTime)
+        return LatexProjectFilesets(allFilesets, dataMapping)
+    }
+
+    private val CACHE_KEY = ProjectCacheService.createKey<LatexProjectFilesets>()
+
+    private suspend fun buildFilesetsSuspend(project: Project): LatexProjectFilesets {
+        return withBackgroundProgress(project, "Building filesets") {
```

Comment:
Right now if you open a non-LaTeX project this task will obscure the background process that is indexing files, maybe because it starts first it will show on top. (if two processes are running, only one will show in the taskbar). That means you can't see the actual progress of the indexing task which is not ideal.
However I'm impressed by how fast building the fileset seems to be - maybe we don't need to show it to the user anymore when it is updating?